### ISSUE TITLE
Freeform incorrectly calls craft\services\Assets::getFolderById() when dynamically creating a directory for file uploads

### ISSUE THREAD

**Issue by tomdavies** [solspace/craft3-freeform#699]
**Describe the bug or issue you're experiencing**

When uploading a file as part of a form submission to a path/folder that must be dynamically created at submission time, a Craft API is incorrectly called, resulting in an error.

**Steps to reproduce**

1. Create a form with an upload control, using a local folder asset volume and `/userUploads/{{ form.handle }}/{{ now|date("Y-m-d H:i:s") }}/` as the `Upload Location Subfolder` value
2. Submit to the form from the front end (I am using a custom template but pretty sure that doesn't make any difference - see below)

<details>
<summary>3. I get this error (click to expand):</summary>

```<pre>Exception 'TypeError' with message 'craft\services\Assets::getFolderById(): Argument #1 ($folderId) must be of type int, craft\models\VolumeFolder given, called in /var/www/html/vendor/solspace/craft-freeform/packages/plugin/src/Services/FilesService.php on line 448' 

in /var/www/html/vendor/craftcms/cms/src/services/Assets.php:401

Stack trace:
#0 /var/www/html/vendor/solspace/craft-freeform/packages/plugin/src/Services/FilesService.php(448): craft\services\Assets-&gt;getFolderById(Object(craft\models\VolumeFolder))
#1 /var/www/html/vendor/solspace/craft-freeform/packages/plugin/src/Services/FilesService.php(83): Solspace\Freeform\Services\FilesService-&gt;getFolder(1, 'userUploads/7aa...', Object(Solspace\Freeform\Form\Types\Regular))
#2 /var/www/html/vendor/solspace/craft-freeform/packages/plugin/src/Fields/FileUploadField.php(132): Solspace\Freeform\Services\FilesService-&gt;uploadFile(Object(Solspace\Freeform\Fields\FileUploadField), Object(Solspace\Freeform\Form\Types\Regular))
#3 /var/www/html/vendor/solspace/craft-freeform/packages/plugin/src/Library/Composer/Components/Form.php(1298): Solspace\Freeform\Fields\FileUploadField-&gt;uploadFile()
#4 /var/www/html/vendor/solspace/craft-freeform/packages/plugin/src/Library/Composer/Components/Form.php(767): Solspace\Freeform\Library\Composer\Components\Form-&gt;validate()
#5 /var/www/html/vendor/solspace/craft-freeform/packages/plugin/src/Controllers/SubmitController.php(52): Solspace\Freeform\Library\Composer\Components\Form-&gt;handleRequest(Object(craft\web\Request))
#6 [internal function]: Solspace\Freeform\Controllers\SubmitController-&gt;actionIndex()
#7 /var/www/html/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#8 /var/www/html/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction-&gt;runWithParams(Array)
#9 /var/www/html/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller-&gt;runAction('', Array)
#10 /var/www/html/vendor/craftcms/cms/src/web/Application.php(302): yii\base\Module-&gt;runAction('freeform/submit', Array)
#11 /var/www/html/vendor/craftcms/cms/src/web/Application.php(627): craft\web\Application-&gt;runAction('freeform/submit', Array)
#12 /var/www/html/vendor/craftcms/cms/src/web/Application.php(281): craft\web\Application-&gt;_processActionRequest(Object(craft\web\Request))
#13 /var/www/html/vendor/yiisoft/yii2/base/Application.php(384): craft\web\Application-&gt;handleRequest(Object(craft\web\Request))
#14 /var/www/html/web/index.php(12): yii\base\Application-&gt;run()
#15 {main}</pre>
```

</details>

Inspecting [`FilesService.php#L445-L449` ](https://github.com/solspace/craft-freeform/blob/904ce98c6a91f711bb4cba416df7649286f84ea5/packages/plugin/src/Services/FilesService.php#L445-L449) the fix looks super easy

Freeform has:

```php
            if (!$folder) {
                $volume = \Craft::$app->getVolumes()->getVolumeById($volumeId);
                $folderId = $assetsService->ensureFolderByFullPathAndVolume($subpath, $volume);
                $folder = $assetsService->getFolderById($folderId);
            }
```

As the error indicates [`craft\services\Assets::ensureFolderByFullPathAndVolume()`](https://docs.craftcms.com/api/v4/craft-services-assets.html#method-ensurefolderbyfullpathandvolume) returns a `VolumeFolder` already, so there is no need to call `craft\services\Assets::getFolderById()`

Looks like you just need:

```php
            if (!$folder) {
                $volume = \Craft::$app->getVolumes()->getVolumeById($volumeId);
                $folder = $assetsService->ensureFolderByFullPathAndVolume($subpath, $volume);
            }
```

**Craft & Plugin Info (please complete the following information):**

- Craft Version: 4.4.11
- Freeform Version: 4.0.25
- Freeform Edition: Pro
- Fresh Install or Upgrade: Fresh

**Comment by kjmartens** [comment]
Sorry for the trouble @tomdavies, and thanks for this. I will note this for our developers to review. 🙂

**Comment by kjmartens** [comment]
This should be resolved now in Freeform 4.1.1. 🙂
